state/account.go

```diff
@@ -93,13 +126,67 @@ func (st *Account) Deserialize(buf []byte) error {
 	return nil
 }
 
+// IsNewbieAccount returns true if the account has not sent any actions
+func (st *Account) IsNewbieAccount() bool {
+	return st.nonce == 0
+}
+
+// SetNonce sets the nonce according to type
+func (st *Account) SetNonce(nonce uint64) error {
+	switch st.accountType {
+	case 1:
+		if nonce != st.nonce {
+			return errors.Wrapf(ErrInvalidNonce, "actual value %d, %d expected", nonce, st.nonce)
+		}
+		st.nonce = nonce + 1
+	case 0:
+		if nonce != st.nonce+1 {
+			return errors.Wrapf(ErrInvalidNonce, "actual value %d, %d expected", nonce, st.nonce+1)
+		}
+		st.nonce = nonce
+	default:
+		return errors.Wrapf(ErrUnknownAccountType, "account type %d", st.accountType)
+	}
+
+	return nil
+}
+
+// PendingNonce returns the pending nonce of the account
+func (st *Account) PendingNonce() uint64 {
+	switch st.accountType {
+	case 1:
+		return st.nonce
+	default: // 0
+		return st.nonce + 1
+	}
+}
+
+// HasSufficientBalance returns true if balance is larger than amount
+func (st *Account) HasSufficientBalance(amount *big.Int) bool {
+	if amount == nil {
+		return true
+	}
+	return amount.Cmp(st.Balance) <= 0
+}
+
 // AddBalance adds balance for account state
-func (st *Account) AddBalance(amount *big.Int) {
-	st.Balance.Add(st.Balance, amount)
+func (st *Account) AddBalance(amount *big.Int) error {
+	if amount == nil || amount.Sign() < 0 {
+		return errors.Wrapf(ErrInvalidAmount, "amount %s shouldn't be negative", amount.String())
+	}
+	if st.Balance != nil {
+		st.Balance = new(big.Int).Add(st.Balance, amount)
```

Comment:
`st.Balance.Add(st.Balance, amount)` would work, same as `SubBalance()` and other places in PR